### ISSUE TITLE
Error 122

### ISSUE THREAD

**Issue by kevfengler227** [chanzuckerberg/shasta#49]
Encountered the following error when running this command on a 2 TB, 32 core Linux machine. Any suggestions?

Thanks,
KF

shasta-Linux-0.1.0 --input subreads.fasta --output SHASTA

2019-Aug-21 23:15:16.174763 LowHash iteration 5 begins.
2019-Aug-22 02:36:28.718041 Terminated after catching a runtime error exception:
Error 122 during ftruncate to size 1987878912

**Comment by paoloczi** [comment]
Please rerun with [Shasta Release 0.2.0](https://github.com/chanzuckerberg/shasta/releases/tag/0.2.0), which was released a couple of days ago. Hopefully that will give us a better indication of what is going on.

For large runs, for best performance and reliability you should use command line options `--memoryMode filesystem --memoryBacking 2M` - see [here](https://chanzuckerberg.github.io/shasta/Performance.html) and [here](https://chanzuckerberg.github.io/shasta/Running.html#MemoryModes) for more information. However using these options requires root access to the machine via `sudo`. If it is impossible for you to obtain root access, it should still be possible to run Shasta, at reduced performance.

Please report on the outcome of running with Release 0.2.0.

**Comment by kevfengler227** [comment]
OK, thanks. I will do that, but with 0.2.0 I get this?

FATAL: kernel too old
Aborted (core dumped)

**Comment by paoloczi** [comment]
What Linux distribution are you on? Please attach the output of `uname -a`.

**Comment by kevfengler227** [comment]
Ah, the machine I was on has Linux 2.6, but I have another with 3.10 which runs 0.2.0. Thanks! I'll give this a try.

**Comment by paoloczi** [comment]
Thanks, let me know how that goes.

Some background information on this:

* For Shasta release 0.1.0, the static executable that was released was built on Ubuntu 16.04. That build should be able to support kernel versions 2.6.32 and newer. 
* For Shasta release 0.2.0, the static executable that was released was built on Ubuntu 18.04. That build should be able to support kernel versions 3.2 and newer.

This means that users on some older systems (for example CentOS  6 - that may be your situation) will not be able to run  Shasta 0.2.0. Given this, for the next release I will consider releasing both versions of the executable.

Building on Ubuntu 16.04 is still supported by the latest Shasta code. So in the meantime, before the next release, I could make a test build of the latest code on Ubuntu 16.04 and give you access to the executable - that would allow you to run the latest version of Shasta on the older system. Let me know if you want me to do that.

**Comment by paoloczi** [comment]
Because of a bug, some data structures used during the LowHash algorithm are always backed by files on disk, regardless of the memory mode options specified. This does not affect the results but lowers performance slightly and increases disk space requirements. Even with this bug, these files are being removed when the LowHash portion of the assembly is complete.

The error message posted at the beginning of this issue (Error 122) was caused by exceeding the user quota of disk space because of these data structures being written to disk.

So to summarize this issue uncovered three separate problems:

* The cryptic message "Error 122". This is fixed in release 0.2.0.
* Release 0.2.0 does not work on some old kernels, but release 0.1.0 did.
* Some LowHash data structures are being written to disk regardless of memory mode options in use.

**Comment by paoloczi** [comment]
I am closing this issue because the cryptic message was already fixed, the LowHash problem was just fixed by the above commit, and for the old kernel problem I created a new issue #50 
 
If new problems arise, please create a new issue rather than appending here.
